File: backend/services/post-ts-koa-mongo-fdm/src/delete-post/impl.ts
import { PostId, UserId } from "../common";
import { Deps } from "../common/utils";
import { Workflow } from ".";
import { ObjectId, WithId } from "mongodb";
import * as db from "../common/utils/db";
import * as runtypes from "runtypes";
import { errors } from "./api";

export class WorkflowImpl extends Workflow {
    private static readonly postHasCreator = runtypes.Record({
        creator: runtypes.InstanceOf(ObjectId),
    });
    async getPostCreator(postId: PostId): Promise<UserId> {
        const oid = db.tryParseId(postId);
        if (oid === undefined) {
            throw errors.postNotFound();
        }
        const post: WithId<unknown> | null = await this.deps.mongoDb.collection(db.posts).findOne({ _id: oid });
        if (post === null) {
            throw this.errors.postNotFound();
        }
        db.validate(WorkflowImpl.postHasCreator, db.posts, post);
        return db.formatId(post.creator) as UserId;
    }
    async deletePost(postId: PostId): Promise<void> {
        const oid = db.tryParseId(postId);
        if (oid === undefined) {
            throw errors.postNotFound();
        }
        await this.deps.mongoDb.collection(db.posts).deleteOne({ _id: oid });
    }

    constructor(private deps: Deps) { super(); }
    readonly errors = errors;
}

File: backend/services/post-ktor-mongo/src/main/kotlin/me/jason5lee/post_ktor_mongo/delete_post/impl.kt
package me.jason5lee.post_ktor_mongo.delete_post

import com.mongodb.client.model.Filters
import com.mongodb.client.model.Projections
import kotlinx.coroutines.reactive.awaitFirstOrNull
import me.jason5lee.post_ktor_mongo.common.PostId
import me.jason5lee.post_ktor_mongo.common.UserId
import me.jason5lee.post_ktor_mongo.common.utils.Db
import me.jason5lee.post_ktor_mongo.common.utils.Deps

class WorkflowImpl(private val deps: Deps) : Workflow(), FailuresImpl {
    override suspend fun getPostCreator(post: PostId): UserId {
        val doc = Db.findById(
            db = deps.mongoDb,
            collection = Db.posts,
            id = Db.toObjectIdOrNull(post.value) ?: throw postNotFound(),
            projection = Projections.include("creator")
        ) ?: throw postNotFound()
        return UserId(Db.formatId(doc.get("creator")))
    }

    override suspend fun deletePost(post: PostId) {
        deps.mongoDb.getCollection(Db.posts)
            .deleteOne(Filters.eq("_id", Db.toObjectIdOrNull(post.value) ?: throw postNotFound()))
            .awaitFirstOrNull()
    }
}
